### ISSUE TITLE
DBSCAN behaviour is different from what is described in the original article.

### ISSUE THREAD

**Issue by iad-ABDUL-RAOUF** [mlpack/mlpack#3339]
Not sure if I should open bug issue or a question issue.
Before using the DBSCAN implementation provided by mlpack, I inspected the code (on master branch) to assert it was clustering as described in the 1996 article [1]. **It seems it does not cluster like DBSCAN should**.

**DBSCAN In mlpack :**
see mlpack/methods/dbscan/dbscan_impl.hpp file.
Using the UnionFind class, it forms clusters of points that can be reach step by step through step of size epsilon.
THEN it checks if each cluster count more than minpts element.

**DBSCAN In the original article :**
See "algorithm 1" in this 2017 article [2] published by the same authors. It describes the original DBSCAN in clearer term.
For each point it looks for its neighbors in an epsilon radius. BEFORE processing the next point, it checks if this neighborhood contains at least minpts element. IF NOT the cluster is not propagated from the current point.

![DBSCAN_algo](https://user-images.githubusercontent.com/71873964/207355716-80025333-6844-4c18-aa64-7994d085aacd.png)

As consequences, the original algorithm is more robust to noisy dataset.
Calling this current mlpack implementation "DBSCAN" is also misleading for users expecting it to actually run the official DBSCAN version.

To fix this issue, I would suggest looking at the scikit-learn implementation (files _dbscan.py and _dbscan_inner.pyx) : 
https://github.com/scikit-learn/scikit-learn/blob/dc580a8ef5ee2a8aea80498388690e2213118efd/sklearn/cluster/_dbscan.py
https://github.com/scikit-learn/scikit-learn/blob/dc580a8ef5ee2a8aea80498388690e2213118efd/sklearn/cluster/_dbscan_inner.pyx 

[1] Ester, Martin, et al. "A density-based algorithm for discovering clusters in large spatial databases with noise." kdd. Vol. 96. No. 34. 1996.
[2] Schubert, Erich, et al. "DBSCAN revisited, revisited: why and how you should (still) use DBSCAN." ACM Transactions on Database Systems (TODS) 42.3 (2017): 1-21.

**Comment by iad-ABDUL-RAOUF** [comment]
Let me complete my post by adding a code example to reproduce the bug.

Could somebody label this issue as a bug ?

```cpp
#include <mlpack/methods/dbscan.hpp>
#include <mlpack/core.hpp>
#include <armadillo>
#include <algorithm>
#include <vector>
#include <iostream>

int main(int argc, char* argv[])
{
    std::vector<std::vector<double>> dataset;
    // cluster 0
    dataset.push_back({0,0});
    dataset.push_back({0.5,0.5});
    dataset.push_back({0.5,-0.5});
    dataset.push_back({1,0});

    // cluster 1
    dataset.push_back({3,0});
    dataset.push_back({3.5,0.5});
    dataset.push_back({3.5,-0.5});
    dataset.push_back({4,0});

    // noise inbetween
    dataset.push_back({2,0});

    // convert std vector into arma matrice
    arma::mat armadataset(dataset[0].size(),dataset.size());
    for (size_t i = 0; i < dataset.size(); ++i){
        for (size_t j = 0; j < dataset[0].size(); ++j) armadataset(j,i) = dataset[i][j];
    }
    std::cout << armadataset << std::endl;
    // 0   0.5000   0.5000   1.0000   3.0000   3.5000   3.5000   4.0000   2.0000
    // 0   0.5000  -0.5000        0        0   0.5000  -0.5000        0        0

    // clustering with mlpack DBSCAN
    double epsilon = 1.1;
    std::size_t minpts = 4;
    using RangeSearchBall = mlpack::RangeSearch<mlpack::EuclideanDistance, arma::mat, mlpack::BallTree>;
    RangeSearchBall rs  = RangeSearchBall();
    mlpack::OrderedPointSelection pointSelector = mlpack::OrderedPointSelection();
    const bool bacthMode = false;
    mlpack::DBSCAN<RangeSearchBall, mlpack::OrderedPointSelection> dbscan(
        epsilon,
        minpts,
        bacthMode,
        rs,
        pointSelector
    );
    arma::Row<size_t> mlpackLabels;
    arma::mat centroids;
    const size_t nclusters = dbscan.Cluster(armadataset, mlpackLabels, centroids);

    // display results
    std::cout << mlpackLabels << std::endl;
    // it displays : 0,0,0,0,0,0,0,0,0 which is wrong.
    // The noisy element has linked the two clusters together
    // The expected output is 0,0,0,0,1,1,1,1,0

    // As a side note, notice that the last noisy element label in between the two clusters should
    // indeed be 0 and not a noise label.
    // This is because this element is in the epsilon-neighborhood of nearby core cluster members.
    // As a consequence, it will get the label of the nearby cluster built first.
    // But it does not have enough neighbors to propagate the cluster label. Cluster 0 and 1 should not
    // be linked together
    return 0;
}
```

**Comment by rcurtin** [comment]
Thanks @iad-ABDUL-RAOUF.  I haven't had the time to look into this closely, but I certainly appreciate the level of detail you've provided here.  Stay tuned and I'll get to this. :+1:

**Comment by AdarshSantoria** [comment]
Hey @rcurtin I worked on the issue and noticed that in dbscan implementation the algorithm is assuming all border points to core points. Due to this some clusters overlap during dbscan clustering. Shall I raise a pr for the implementation of border points ?

**Comment by AdarshSantoria** [comment]
Hey @rcurtin I worked on the issue and noticed that in dbscan implementation the algorithm is assuming all border points to core points. Due to this some clusters overlap during dbscan clustering. Shall I add a pr for the implementation of border points ?

**Comment by SuvarshaChennareddy** [comment]
Mind if I try taking a look at this?

**Comment by rcurtin** [comment]
#3346 is merged, so this one is fixed. :+1:
